include/swift/Basic/Statistic.h

```diff
@@ -110,20 +109,36 @@ class UnifiedStatsReporter {
   // active at all. Reduces redundant machinery.
   class RecursionSafeTimers;
 
+  // We also keep a few banks of optional hierarchical profilers for times and
+  // statistics, activated with -profile-stats-events and
+  // -profile-stats-entities, which are part way between the detail level of the
+  // aggregate statistic JSON files and the fine-grained CSV traces. Naturally
+  // these are written in yet a different file format: the input format for
+  // flamegraphs.
+  struct StatsProfilers;
+
 private:
   bool currentProcessExitStatusSet;
   int currentProcessExitStatus;
   SmallString<128> StatsFilename;
   SmallString<128> TraceFilename;
+  SmallString<128> ProfileDirname;
   llvm::TimeRecord StartedTime;
+
+  // This is unique_ptr because NamedRegionTimer is non-copy-constructable.
   std::unique_ptr<llvm::NamedRegionTimer> Timer;
+
   SourceManager *SourceMgr;
   clang::SourceManager *ClangSourceMgr;
-  std::unique_ptr<AlwaysOnDriverCounters> DriverCounters;
-  std::unique_ptr<AlwaysOnFrontendCounters> FrontendCounters;
-  std::unique_ptr<AlwaysOnFrontendCounters> LastTracedFrontendCounters;
-  std::vector<FrontendStatsEvent> FrontendStatsEvents;
+  llvm::Optional<AlwaysOnDriverCounters> DriverCounters;
+  llvm::Optional<AlwaysOnFrontendCounters> FrontendCounters;
+  llvm::Optional<AlwaysOnFrontendCounters> LastTracedFrontendCounters;
+  llvm::Optional<std::vector<FrontendStatsEvent>> FrontendStatsEvents;
```

Comment:
Nitpick: No need for the `llvm::` prefix if you import swift/Basic/LLVM.h

Reply:
I never knew! Nice.